async/component.go

```diff
@@ -38,36 +41,102 @@ type Component struct {
 	retryWait    time.Duration
 }
 
-// New returns a new async component. The default behavior is to return a error of failure.
-// Use options to change the default behavior.
-func New(name string, p ProcessorFunc, cf ConsumerFactory, oo ...OptionFunc) (*Component, error) {
+// Builder gathers all required properties in order to construct a component
+type Builder struct {
+	errors       []error
+	name         string
+	proc         ProcessorFunc
+	failStrategy FailStrategy
+	cf           ConsumerFactory
+	retries      uint
+	retryWait    time.Duration
+}
 
+// New initializes a new builder for a component with the given name
+// by default the failStrategy will be NackExitStrategy
+func New(name string) *Builder {
+	var errs []error
 	if name == "" {
-		return nil, errors.New("name is required")
+		errs = append(errs, errors.New("name is required"))
 	}
+	return &Builder{
+		name:   name,
+		errors: errs,
+	}
+}
 
-	if p == nil {
-		return nil, errors.New("work processor is required")
+// WithProcessor sets the processing logic for the component
+// it will append an error to the builder if the processor is nil
+func (cb *Builder) WithProcessor(proc ProcessorFunc) *Builder {
+	if proc == nil {
+		cb.errors = append(cb.errors, errors.New("work processor is required"))
+	} else {
+		log.Infof(propSetMSG, "processor func", cb.name)
+		cb.proc = proc
 	}
+	return cb
+}
 
+// WithConsumerFactory defines the consumer factory to be used in order to create the appropriate consumer
+// it will append an error to the builder if the factory is nil
+func (cb *Builder) WithConsumerFactory(cf ConsumerFactory) *Builder {
 	if cf == nil {
-		return nil, errors.New("consumer is required")
+		cb.errors = append(cb.errors, errors.New("consumer is required"))
+	} else {
+		log.Infof(propSetMSG, "consumer factory", cb.name)
+		cb.cf = cf
 	}
+	return cb
+}
 
-	c := &Component{
-		name:         name,
-		proc:         p,
-		cf:           cf,
-		failStrategy: NackExitStrategy,
-		retries:      0,
-		retryWait:    0,
+// WithFailureStrategy defines the failure strategy to be used
+// default value is NackExitStrategy
+// it will append an error to the builder if the strategy is not one of the pre-defined ones
+func (cb *Builder) WithFailureStrategy(fs FailStrategy) *Builder {
+	if _, ok := failStrategyValues[fs]; ok {
```

Comment:
As I understand it, we are using the map to check if we support the failurestrategy or if and invalid value has been passed in.
If that is the reason we should do a simple switch statement to see if the value is acceptable or not, and avoid having an map allocated